src/stmt.cpp

```diff
@@ -0,0 +1,272 @@
+
+void LoopSeqStmt::emit(std::ostream &stream, std::string offset) {
+    stream << offset << "/* LoopSeq " << std::to_string(loops.size()) << " */\n";
+
+    for (const auto &loop : loops) {
+        loop.first->emitPrefix(stream, offset);
+        loop.first->emitHeader(stream, offset);
+        loop.second->emit(stream, offset);
+        loop.first->emitSuffix(stream, offset);
+    }
+}
+
+std::shared_ptr<LoopSeqStmt> LoopSeqStmt::generateStructure(std::shared_ptr<GenCtx> ctx) {
+    auto gen_pol = ctx->getGenPolicy();
+    size_t loop_num = rand_val_gen->getRandId(gen_pol->loop_seq_num_distr);
+
+    auto new_loop_seq = std::make_shared<LoopSeqStmt>();
+    auto new_ctx = std::make_shared<GenCtx>(*ctx);
+    //TODO: is it the right place to do it?
+    new_ctx->incLoopDepth(1);
+    for(size_t i = 0; i < loop_num; ++i) {
+        auto new_loop_head = std::make_shared<LoopHead>();
+        size_t iter_num = rand_val_gen->getRandId(gen_pol->iters_num_distr);
+        for (size_t iter_idx = 0; iter_idx < iter_num; ++iter_idx)
+            new_loop_head->addIterator(Iterator::create(ctx));
+        auto new_loop_body = ScopeStmt::generateStructure(new_ctx);
+        new_loop_seq->addLoop(std::make_pair(new_loop_head, new_loop_body));
+    }
+    return new_loop_seq;
+}
+
+void LoopSeqStmt::populate(std::shared_ptr<PopulateCtx> ctx) {
+    for (auto &loop : loops) {
+        auto loop_head = loop.first;
+        if (loop_head->getPrefix().use_count() != 0)
+            loop_head->getPrefix()->populate(ctx);
+        auto new_ctx = std::make_shared<PopulateCtx>(ctx);
+        new_ctx->incLoopDepth(1);
+        new_ctx->getLocalSymTable()->addIters(loop_head->getIterators());
+        loop.second->populate(new_ctx);
+        new_ctx->decLoopDepth(1);
+        new_ctx->getLocalSymTable()->deleteLastIters();
+        if (loop_head->getSuffix().use_count() != 0)
+            loop_head->getSuffix()->populate(ctx);
+    }
+}
+
+void LoopNestStmt::emit(std::ostream &stream, std::string offset) {
+    stream << offset << "/* LoopNest " << std::to_string(loops.size()) << " */\n";
+
+    std::string new_offset = offset;
+    for (const auto &loop : loops) {
+        loop->emitPrefix(stream, new_offset);
+        loop->emitHeader(stream, new_offset);
+        stream << "{\n";
+        new_offset += "    ";
+    }
+
+    body->emit(stream, new_offset);
+    new_offset.erase(new_offset.size() - 4, 4);
+
+    for (const auto &loop : loops) {
+        stream << new_offset << "} \n";
+        loop->emitSuffix(stream, new_offset);
+        new_offset.erase(new_offset.size() - 4, 4);
+    }
+}
+
+std::shared_ptr<LoopNestStmt> LoopNestStmt::generateStructure(std::shared_ptr<GenCtx> ctx) {
+    auto gen_pol = ctx->getGenPolicy();
+    size_t nest_depth = rand_val_gen->getRandId(gen_pol->loop_nest_depth_distr);
+
+    // We need to limit a maximal loop depth that we can generate
+    nest_depth = std::min(gen_pol->loop_depth_limit - ctx->getLoopDepth(), nest_depth);
```

Comment:
VS2017 build:
C:\projects\yarpgen\src\stmt.cpp(230): error C2039: 'min': is not a member of 'std'